docs/architecture/adr-076-sign-mode-unified.md

```diff
@@ -0,0 +1,327 @@
+
+The metadata digest is a compact representation of the complete metadata set. The hash of this digest serves as the metadata hash, which is included in the transaction's sign doc. Here's the proposed structure for the Hash type and the MetadataDigest in Go:
+
+```go
+type Hash [32]byte
+
+type MetadataDigest struct {
+    TreeRootHash     Hash   // The Merkle root hash of the metadata tree
+    SpecVersion      uint32 // A version number for the metadata specification
+    Props            map[string]string // A map of properties for the metadata
+}
+```
+
+The field `Props` is a map of properties for the metadata, it is used to store additional information about the metadata that are chain-specific. For example, it can be used to store the token symbols and decimals.
+
+### Metadata merkle tree
+
+The metadata merkle tree is a crucial component of the `SIGN_MODE_UNIFIED` design. It allows for efficient verification of metadata integrity without requiring offline signers to store the full metadata. The tree is constructed as a complete binary merkle tree using `blake3` as the hashing function.
+
+#### Construction Process
+
+1. **Prepare the leaves**: The initial data (leaves of the merkle tree) are the type information sorted by their unique identifiers. For `Enumeration` types, variants are sorted using their `index`.
+
+2. **Sort the leaves**: The leaves are sorted to ensure the tree root is deterministic.
+
+3. **Build the tree**: The tree is built from the bottom up, combining pairs of hashes to form parent nodes until a single root hash is obtained.
+
+4. **Handle empty trees**: If there are no nodes left in the list (i.e., the initial data set was empty), an all-zeros hash `[32]byte{0}` is used to represent the empty tree.
+
+The resulting `merkleTreeRoot` is the last node left in the list of nodes, representing the root of the entire merkle tree.
+
+#### Purpose and Benefits
+
+* **Efficient Verification**: The merkle tree structure allows for efficient verification of metadata integrity. Only the necessary chunks and their corresponding proofs need to be provided to offline wallets for transaction decoding.
+
+* **Reduced Storage Requirements**: Offline signers don't need to store the full metadata. They can verify the integrity of the metadata using just the root hash and the relevant chunks.
+
+* **Security**: The use of cryptographic hashes in the tree structure ensures the integrity of the metadata, making it tamper-evident.
+
+### Encoding
+
+`SIGN_MODE_UNIFIED` proposes using CBOR (Concise Binary Object Representation) for transaction encoding, offering several critical advantages:
+
+* **Deterministic Encoding**: CBOR's canonical form ensures consistent merkleization of transaction data, which is essential for verification and proof generation.
+* **Human-readable Diagnostic Notation**: While CBOR is binary, it provides a diagnostic notation format that maintains transparency and reduces blind signing risks.
```

Comment:
Hey, thanks for pointing that out, true!
We are proposing CBOR because it enables metadata merkleization, a central feature of this new sign mode. While CBOR is not human-readable in its raw form, any UI can easily decode it to present the data to users. Blind signing would only occur if a wallet implements a "lazy" UI that doesn't decode the CBOR string, which would likely make such a wallet less appealing to users.

This approach doesn't represent a step backwards in terms of verifiability, because:
- CBOR transactions can always be decoded for human readability. We rely on widely-tested codec libraries for this, just as we do with other formats.
- Since the transaction contains the metadata's root hash, the signer can verify that the decoded transaction chunks comply with the metadata sent by the node. This provides a "first-order" verification.
- Finally, since the sign doc contains the root hash (meaning the user is signing both `tx_blob + metadata_root_hash`), validators can verify that the transaction was displayed (and signed) using the consensual metadata tree, providing a "second-order" verification.

Reply:
> We are proposing CBOR because it enables metadata merkleization, a central feature of this new sign mode.

What specifically about CBOR enables metadata merkleization as opposed to say JSON? It's simply an encoding format.

Reply:
The original question here was about human readability. Given the amount of detail that the previous spec for textual put into it human readability (or even amino JSON), I find "CBOR transactions can always be decoded for human readability" to be a pretty underwhelming answer.